src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,363 @@
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddQueryTime(o)))
+            {
+                header = this.CheckCache(blockHeight);
+
+                if (header == null)
+                {
+                    // Check the repository (DBreeze).
+                    header = await this.provenBlockHeaderRepository.GetAsync(blockHeight).ConfigureAwait(false);
+
+                    this.AddToCache(blockHeight, header);
+                }
+            }
+
+            return header;
+        }
+
+        /// <inheritdoc />
+        public async Task<List<ProvenBlockHeader>> GetAsync(int fromBlockHeight, int toBlockHeight)
+        {
+            var headersInCache = new List<ProvenBlockHeader>();
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddQueryTime(o)))
+            {
+                int heightKey = fromBlockHeight;
+
+                ProvenBlockHeader header = null;
+
+                var headersNotInCache = new List<int>();
+
+                do
+                {
+                    header = this.CheckCache(heightKey);
+
+                    if (header != null)
+                        headersInCache.Add(header);
+                    else
+                        headersNotInCache.Add(heightKey);
+
+                    heightKey++;
+
+                } while (heightKey <= toBlockHeight);
+
+                // Try and get items from the repository if not found in the store cache.
+                if (headersNotInCache.Count > 0)
+                {
+                    var repositoryHeaders = new List<ProvenBlockHeader>();
+
+                    foreach(int headerNotInCache in headersNotInCache)
+                    {
+                        var repositoryHeader = new ProvenBlockHeader();
+
+                        repositoryHeader = await this.provenBlockHeaderRepository.GetAsync(headerNotInCache).ConfigureAwait(false);
+
+                        if (repositoryHeader != null)
+                        {
+                            repositoryHeaders.Add(repositoryHeader);
+
+                            this.AddToCache(headerNotInCache, repositoryHeader);
+                        }
+                    }
+
+                    if (repositoryHeaders.Count > 0)
+                        headersInCache.AddRange(repositoryHeaders);
+                }
+            }
+
+            return headersInCache;
+        }
+
+        /// <inheritdoc />
+        public void AddToPendingBatch(ProvenBlockHeader provenBlockHeader, HashHeightPair newTip)
+        {
+            this.PendingBatch.AddOrUpdate(newTip.Height, provenBlockHeader, (key, value) => { return provenBlockHeader; });
+
+            this.Cache.AddOrUpdate(newTip.Height, provenBlockHeader, (key, value) => { return provenBlockHeader; });
+
+            this.pendingTipHashHeight = newTip;
+        }
+
+        /// <summary> Saves pending <see cref="ProvenBlockHeader"/> items to the <see cref="ProvenBlockHeaderRepository"/>.</summary>
```

Comment:
Instead of concrete type use interface when mentioning repository

Reply:
ok